contractsV5/KyberHintParser.sol

```diff
@@ -0,0 +1,170 @@
+    )
+        internal
+        view
+    {
+        bytes memory opcode = hint.slice(indexToContinueFrom,1);
+        bytes32 opcodeHash = keccak256(opcode);
+        ReservesHint memory reserves;
+        if (isTokenToEth) {
+            reserves = tradeHint.tokenToEthReserves;
+        } else {
+            reserves = tradeHint.ethToTokenReserves;
+        }
+
+        indexToContinueFrom += 1;
+        if (opcodeHash == keccak256(END_OPCODE)) {
+            return;
+        } else if (opcodeHash == keccak256(SEPARATOR)) {
+            decodeOperation(hint, tradeHint, indexToContinueFrom, false);
+        } else if ((opcodeHash == keccak256(MASK_IN_OPCODE) || opcodeHash == keccak256(MASK_OUT_OPCODE))) {
+            reserves.isSplit = false;
+            (reserves.addresses, reserves.splitValuesBps, indexToContinueFrom) = decodeReservesFromHint(reserves.isSplit, hint, indexToContinueFrom);
+            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth);
+        } else if (opcodeHash == keccak256(SPLIT_TRADE_OPCODE)) {
+            reserves.isSplit = true;
+            (reserves.addresses, reserves.splitValuesBps, indexToContinueFrom) = decodeReservesFromHint(reserves.isSplit, hint, indexToContinueFrom);
+            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth);
+        } else {
+            revert("Invalid hint opcode");
+        }
+    }
+
+    function decodeReservesFromHint(bool isTokenSplit, bytes memory hint, uint indexToContinueFrom)
+        public
+        pure
+        returns
+        (address[] memory, uint[] memory, uint)
+    {
+       uint reservesLength = hint.toUint8(indexToContinueFrom);
+       uint bpsSoFar;
+       uint[] memory splitValuesBps;
+
+       address[] memory reserves = new address[](reservesLength);
+       if (isTokenSplit) {
+           splitValuesBps = new uint[](reservesLength);
+       } else {
+           splitValuesBps = new uint[](1);
+           splitValuesBps[0] = BPS;
+       }
+       indexToContinueFrom++;
+
+       for (uint i = 0; i < reservesLength; i++) {
+           reserves[i] = hint.toAddress(indexToContinueFrom);
+           indexToContinueFrom += 20;
+           if (isTokenSplit) {
+               splitValuesBps[i] = uint(hint.toUint16(indexToContinueFrom));
+               bpsSoFar += splitValuesBps[i];
+               indexToContinueFrom += 2;
+           }
+       }
+       require((bpsSoFar == BPS) || (bpsSoFar == 0), "bps do not sum to 10000");
+       return (reserves, splitValuesBps, indexToContinueFrom);
+    }
+
+    function getBytes(bytes memory _hint, uint _start, uint _length) public pure returns (bytes memory) {
+        return _hint.slice(_start,_length);
+    }
+
+    function getSingleByte(bytes memory _hint, uint _index) public pure returns (bytes memory) {
+        return abi.encodePacked(_hint[_index]);
+    }
+
+    function buildHint(
+        bytes memory e2tOpcode,
+        address[] memory e2tReserves,
```

Comment:
here should be changed to reserve IDs
so uint array.
or maybe uint4 array.
or bytes4 array.
not sure what will be best for parsing and building made easy...

Reply:
Currently using bytes3 array.